_posts/2023-10-02-kafka-consumer-client-essentials.md

```diff
@@ -0,0 +1,457 @@
+
+Here's a breakdown of retriable and non-retriable errors that the client handles:
+
+**Non-retriable errors caught by the consumer application**
+
+* `InterruptedException`: This error occurs when the current thread is interrupted while paused. 
+Interruption typically happens during consumer shutdown or when stopping its operation. 
+The exception is rethrown as a `RuntimeException`, which ultimately terminates the consumer.
+
+* `IllegalArgumentException`: This error is thrown when the consumer receives invalid or inappropriate arguments. 
+For instance, it can be triggered if essential details like the topic are missing.
+
+* `UnsupportedOperationException`: This error is raised when an operation is not supported or when a method is not implemented as expected. 
+For instance, it can be triggered if you attempt to use an unsupported consumer configuration, leading to a runtime exception.
+
+**Retriable errors caught by the consumer application**
+
+`RetriableException`: This type of error is thrown for any exception that implements the `RetriableException` interface, as provided by the Kafka client library.
+
+`OffsetOutOfRangeException`: This error is thrown when the consumer attempts to seek to an invalid offset for a partition, typically when the offset is outside the valid range of offsets for that partition. To recover, the consumer seeks to the end of the partition to commit the offset synchronously (`commitSync`).
+
+`NoOffsetForPartitionException`: This error is thrown when there is no committed offset for a partition, and auto-reset policy is not enabled, or the requested offset is invalid. To recover, the consumer seeks to the end of the partition to commit the offset synchronously (`commitSync`).
+
+`RebalanceInProgressException`: This error is thrown during a consumer group rebalance when partitions are being assigned. Offset commits cannot be completed when the consumer is undergoing a rebalance. The consumer can wait for the rebalance to complete and then continue processing.
+
+## Tuning your consumer
+
+The consumer application provided in this post serves as a starting point, but you can fine-tune it to align with your specific requirements. 
+When tuning a consumer application, consider the following aspects, as they significantly impact its performance and behavior:
+
+1. Scaling: Consumer groups enable parallel processing of messages by distributing the load across multiple consumers. Properly configuring consumer groups can enhance scalability and throughput.
+2. Message ordering: A consumer observes messages in a single partition in the same order that they were committed to the broker, which means that Kafka only provides ordering guarantees for messages in a single partition. If you want a strict ordering of messages from one topic, use one partition per consumer.
```

Comment:
the above is wrong or I didn't get what you meant.
Saying `use one partition per consumer` doesn't guarantee ordering of messages across the topic. The topic could still have more partitions, each partition assigned to one consumer but ordering not guaranteed.
Maybe you meant to use a topic with just one partition? In this case the order is guaranteed of course, and you will have also just one consumer (per consumer group) reading from that topic.

Reply:
Thanks. Let me see if I have addressed this succinctly:

> 2. Message ordering: A consumer observes messages in a single partition in the same order that they were committed to the broker, which means that Kafka only provides ordering guarantees for messages in a single partition. 
> If absolute ordering within a topic is important, use a single-partition topic. 
> While a single-partition topic allows only one consumer instance, you can leverage a pool of worker threads to create a multi-threaded consumer for more efficient processing.
> When processing messages for specific entities (like users), you can use the ID (like user ID) as the message key and direct all messages to a single partition.
> This way, you can maintain message ordering for events specific to individual entities. 
> If a new entity is created, you can create a new topic and migrate the data.

I'm not entirely sure of that last line and if that would always be the case?

Reply:
> I'm not entirely sure of that last line and if that would always be the case?

Well, a new entity will land on an existing partition. In this case the consumer must be able to process the events based on the entity id. If the requirement is 1 entity = 1 partition, then you are correct, but it is quite unconvenient. I would live that out.